modules/3d/src/spc.cpp

```diff
@@ -0,0 +1,244 @@
+        int index;
+        float distance;
+        dist_node(int i, float d) : index(i), distance(d) {}
+        bool operator<(const dist_node& other) const { return distance > other.distance; }
+    };
+
+    cv::parallel_for_(cv::Range(0, num_faces), [&](const cv::Range& range) {
+        for (int i = range.start; i < range.end; ++i) {
+            std::vector<bool> visited(num_faces, false);
+            std::priority_queue<dist_node> pq;
+            pq.emplace(i, 0.f);
+
+            while (!pq.empty()) {
+                dist_node current = pq.top();
+                pq.pop();
+
+                if (visited[current.index])
+                    continue;
+
+                visited[current.index] = true;
+
+                for (int neighbor = 0; neighbor < num_faces; ++neighbor) {
+                    float weight = in.at<float>(current.index, neighbor);
+                    if (weight > 0 && !visited[neighbor]) {
+                        float distance = current.distance + weight;
+                        if (distance < out.at<float>(i, neighbor)) {
+                            out.at<float>(i, neighbor) = distance;
+                            pq.emplace(neighbor, distance);
+                        }
+                    }
+                }
+            }
+        }
+    });
+
+    out.setTo(0, out == std::numeric_limits<float>::infinity());
+
+    Scalar sum = cv::sum(out);
+    auto sum_val = static_cast<float>(sum[0]);
+    float sigma = sum_val / static_cast<float>(num_faces * num_faces);
+    sigma = 2.0f * sigma * sigma;
+
+    // Gaussian Kernel
+    out.forEach<float>([sigma](float& elem, const int* position) -> void {
+        CV_UNUSED(position);
+        elem = (elem == 0.f) ? 0.f : exp(-elem / sigma);
+    });
+
+    // set diagonal elements to 1
+    for (int i = 0; i < num_faces; ++i)
+        out.at<float>(i, i) = 1.f;
+}
+
+void SpectralCluster::Impl::getAdjacentDistanceMat(Mat &out, const std::vector<Point3f> &vertices,
+                                             const std::vector<std::vector<int32_t>> &indices) const {
+    int num_faces = int(indices.size());
+    int num_adjacency = 0;
+    struct PairHash {
+        std::size_t operator() (const std::pair<int, int>& pair) const {
+            return std::hash<int>{}(pair.first) ^ std::hash<int>{}(pair.second);
+        }
+    };
+    std::unordered_map<std::pair<int, int>, std::pair<int, int>, PairHash> map;
+    map.reserve(num_faces * 3);
+    for (int i = 0; i < num_faces; ++i) {
+        const std::vector<int32_t> &face = indices[i];
+        for (int j = 0; j < 3; ++j) {
+            int v1 = face[j], v2 = face[(j + 1) % 3];
+            if (v1 > v2) std::swap(v1, v2);
+            auto it = map.find({v1, v2});
+            if (it != map.end())
+                it->second.second = i;
+            else
+                map[{v1, v2}] = std::make_pair(i, -1);
+        }
+    }
+    // init
+    Mat g_dist_mat = Mat::zeros(num_faces, num_faces, CV_32F);
```

Comment:
This may eat all available memory on some middle-sized models.
Please consider using different data structure or limit max face number to some realistic number.

Reply:
Hi Savor, I did find this problem when I was implementing this algorithm. Since the spectral clustering algorithm needs to calculate the affinity matrix (Dijkstra algorithm) based on the relationship between each face, it will generate a matrix of size `num_face * num_face`, which takes up too much memory. I have not thought of a better solution to avoid this problem. The possible method I can think of now is to use an algorithm with lower space complexity instead of spectral clustering to implement mesh clustering. I would like to ask if you have any better suggestions? Thanks!

Reply:
@starga2er777 Hi, I don't have any good solution on my mind.
If this matrix is sparse enough you can consider replacing it by some kind of sparse structure like `std::unordered_map<std::pair<int, int>, float>` but I am not sure that this will work.
For now I propose to limit the maximum size of mesh by several thousands faces based on an explicit memory limit. You can specify a constant in your code that shows how much memory we want to fit in and calculate the maximum mesh size from it.